### PR TITLE
[EOC-168] Demo with demo users (changes disappear after logout)

### PR DESCRIPTION
**Pull request by alemikolo**
1. Demo user has an account not connected with Google.
2. I used a passport local strategy (for demo purposes only) to omit Google authentication, and store demo user in request and session.
3. When the demo button is clicked:
    - the request is sent to the backend,
    - demo user is taken from DB,
    - demo user's data is saved to DB with new id
    - demo data is seeded for the new saved demo user,
    - new demo user is sent to the client (I didn't change our current approach and send user via cookie).
4. Demo users can't see normal users' data, can't invite other users to their lists or cohorts.
5. There could be more than one demo user - their accounts are separated,they can try the app independently.
6. Users can't see and invite demo users and can't see their changes.
7. I change passport serialize to store in session only userId not all user's data.
8. Demo user's seeded data and changes are removed on logout
9. On logout, all changes made by the demo user are removed, but logout action is performed only when this user click log out button. I tried to invoke logout action on window 'onbeforeunload' event but then on every page refresh user was logged out and redirected to the login page, so I removed this functionality.
10. I wrote script clearDemoData.js and I added it the package.json to remove demo data manually.
11. To make this functionality working we have to:
- Add DEMO_USER_ID and DEMO MODE_ID constants to .env file. Both should be Hexadecimal Strings with 24 characters. They will be used to generate and remove data for demo purposes.
- Remove index from the "idFromProvider" field in the users collection in DB:
- using MongoDB Compass
- using mongo shell:
```
    mongo --shell
    use eoc
    db.users.getIndexes()
    db.dropIndex("<field name displayed by the above command>")
```

chrome-capture

### CODE REVIEW

File: client/src/common/components/Preloader.scss [hunk 1]
```diff
@@ -16,7 +16,17 @@
   }
 
   &--light {
-    background-color: $dark-orange;
+    // background-color: $dark-orange;
```
Comment (marekrozmus) [inline client/src/common/components/Preloader.scss]: commented out code

File: client/src/modules/authorization/AuthBox.jsx [hunk 2]
```diff
@@ -35,19 +53,41 @@ class AuthBox extends PureComponent {
...
+                    'disabled-google-button': !isCookieSet || pending
+                  })}
+                  onClick={() => this.setState({ pending: true })}
+                  href="/auth/google"
+                  tabIndex={!isCookieSet ? '-1' : '1'}
+                >
+                  <img
+                    alt="Sign in with Google"
+                    className="google-button__img"
+                    src="client/src/assets/images/google-btn.png"
```
Comment (marekrozmus) [inline client/src/modules/authorization/AuthBox.jsx]: remember to merge this correctly

File: server/app.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ const app = express();
 
 // Set up mongodb connection
 const dbUrl = DB_URL;
-mongoose.connect(dbUrl);
+mongoose.connect(`${dbUrl}`);
```
Comment (marekrozmus) [inline server/app.js]: ?

File: server/seed/demoSeed/seedCohorts.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/* eslint-disable no-await-in-loop */
+const Cohort = require('../../models/cohort.model');
+// require('dotenv').config();
```
Comment (marekrozmus) [inline server/seed/demoSeed/seedCohorts.js]: commented out code

File: server/seed/demoSeed/seedComments.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable no-await-in-loop */
+const Comment = require('../../models/comment.model');
+// require('dotenv').config();
```
Comment (marekrozmus) [inline server/seed/demoSeed/seedComments.js]: commented out code

File: server/seed/demoSeed/seedLists.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/* eslint-disable no-await-in-loop */
+const List = require('../../models/list.model');
+// require('dotenv').config();
```
Comment (marekrozmus) [inline server/seed/demoSeed/seedLists.js]: commented out code

File: server/seed/demoSeed/seedUsers.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-await-in-loop */
+const User = require('../../models/user.model');
+// require('dotenv').config();
```
Comment (marekrozmus) [inline server/seed/demoSeed/seedUsers.js]: commented out code

File: client/src/modules/authorization/model/reducer.js [hunk 8]
```diff
@@ -4,7 +4,9 @@ const currentUser = (state = null, action) => {
   switch (action.type) {
     case AuthorizationActionTypes.SET_CURRENT_USER_SUCCESS:
       return action.payload;
-    case AuthorizationActionTypes.LOGOUT_USER_SUCCESS:
+    case AuthorizationActionTypes.LOGIN_SUCCESS:
```
Comment (marekrozmus) [inline client/src/modules/authorization/model/reducer.js]: Could be just moved to line 6 and then both cases will return same as now but without additional code duplication

File: server/common/utils/userUtils.js [hunk 9]
```diff
@@ -29,7 +34,40 @@ const extractUserProfile = (profile, accessToken) => {
   };
 };
 
+const removeDemoUserData = async id => {
```
Comment (marekrozmus) [inline server/common/utils/userUtils.js]: why using async/await? Are they used anywhere in project? As far as I know Promises were used.

File: server/config/auth.js [hunk 10]
```diff
@@ -21,12 +24,43 @@ passport.use(
...
+    User.findOne({
+      idFromProvider: process.env.DEMO_USER_ID_FROM_PROVIDER
+    })
+      .lean()
+      .exec()
+      .then(user => {
+        const { _id, ...rest } = user;
+        return new User({ ...rest }).save();
+      })
+      .then(async newUser => {
```
Comment (marekrozmus) [inline server/config/auth.js]: please do not mix promises with asyncs

File: server/common/utils/userUtils.js [hunk 11]
```diff
@@ -29,7 +34,36 @@ const extractUserProfile = (profile, accessToken) => {
   };
 };
 
+const removeDemoUserData = id =>
+  List.find(
+    {
+      $or: [{ ownerIds: id }, { memberIds: id }, { viewersIds: id }]
```
Comment (adamklepacz) [inline server/common/utils/userUtils.js]: should we check all ids here ?? 
not only ownerIds??

File: server/config/auth.js [hunk 12]
```diff
@@ -21,12 +24,42 @@ passport.use(
   )
 );
 
+// Use LocalStrategy for demo purposes
+passport.use(
+  new LocalStrategy((username, password, done) => {
+    let newUser;
```
Comment (adamklepacz) [inline server/config/auth.js]: add line break

File: server/config/auth.js [hunk 13]
```diff
@@ -21,12 +24,42 @@ passport.use(
...
+    User.findOne({
+      idFromProvider: process.env.DEMO_USER_ID_FROM_PROVIDER
```
Comment (adamklepacz) [inline server/config/auth.js]: destructure this data

File: server/config/auth.js [hunk 14]
```diff
@@ -21,12 +24,42 @@ passport.use(
...
+    })
+      .lean()
+      .exec()
+      .then(user => {
+        const { _id, ...rest } = user;
```
Comment (adamklepacz) [inline server/config/auth.js]: add line break

File: server/config/auth.js [hunk 15]
```diff
@@ -21,12 +24,42 @@ passport.use(
...
+        seedDemoData(newUser._id);
+      })
+      .then(() => done(null, newUser))
+      .catch(err => done(null, false, { message: err.message }));
+  })
+);
+
 passport.serializeUser((user, done) => {
-  done(null, user);
+  done(null, user._id);
```
Comment (adamklepacz) [inline server/config/auth.js]: destructure this data

File: server/controllers/cohort.js [hunk 16]
```diff
@@ -431,14 +431,20 @@ const removeOwnerRole = (req, resp) => {
...
-    user: { _id: userId }
+    user: { _id: userId, idFromProvider }
   } = req;
   const { id: cohortId } = req.params;
   const { email } = req.body;
   let currentCohort;
   let newMember;
   const sanitizedCohortId = sanitize(cohortId);
 
+  if (idFromProvider === process.env.DEMO_USER_ID_FROM_PROVIDER) {
```
Comment (adamklepacz) [inline server/controllers/cohort.js]: destructure this data above if

File: server/controllers/cohort.js [hunk 17]
```diff
@@ -452,7 +458,10 @@ const addMember = (req, resp) => {
       return User.findOne({ email: sanitize(email) }).exec();
     })
     .then(user => {
-      if (!user) {
+      if (
+        !user ||
+        user.idFromProvider === process.env.DEMO_USER_ID_FROM_PROVIDER
```
Comment (adamklepacz) [inline server/controllers/cohort.js]: destructure this data

File: server/controllers/list.js [hunk 18]
```diff
@@ -751,14 +751,20 @@ const removeMemberRole = (req, resp) => {
...
-    user: { _id: currentUserId }
+    user: { _id: currentUserId, idFromProvider }
   } = req;
   const { id: listId } = req.params;
   const { email } = req.body;
   let list;
   let user;
   let cohortMembers = [];
 
+  if (idFromProvider === process.env.DEMO_USER_ID_FROM_PROVIDER) {
```
Comment (adamklepacz) [inline server/controllers/list.js]: destrucutre this data here and everywhere from process

File: server/controllers/userAuth.js [hunk 19]
```diff
@@ -1,27 +1,29 @@
...
-    'user',
-    JSON.stringify({
-      avatarUrl: req.user.avatarUrl,
-      id: req.user.id,
-      name: req.user.displayName
-    })
-  );
-  res.redirect('/');
+const sendUser = (req, resp) => {
+  const { avatarUrl, _id: id, displayName: name } = req.user;
```
Comment (adamklepacz) [inline server/controllers/userAuth.js]: add break line

File: server/controllers/userAuth.js [hunk 20]
```diff
@@ -1,27 +1,29 @@
...
-    res.redirect('/');
+    resp.clearCookie('connect.sid');
+    resp.clearCookie('user');
+    resp.clearCookie('demo');
+    resp.redirect('/');
   });
 };
 
+const sendDemoUser = (req, resp) => {
+  const { avatarUrl, _id: id, displayName: name } = req.user;
```
Comment (adamklepacz) [inline server/controllers/userAuth.js]: add break line

File: server/middleware/removeDemoUserChanges.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
+const { removeDemoUserData } = require('../common/utils/userUtils');
+
+const removeDemoUserChanges = async (req, res, next) => {
```
Comment (adamklepacz) [inline server/middleware/removeDemoUserChanges.js]: is this asyncy needed here? 

you don't have await anywhere inside function

File: server/middleware/removeDemoUserChanges.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  if (!req.user) {
```
Comment (adamklepacz) [inline server/middleware/removeDemoUserChanges.js]: destructure user from req

File: server/seed/demoSeed/generateComments.js [hunk 23]
```diff
@@ -0,0 +1,25 @@
+const mongoose = require('mongoose');
+
+const {
+  Types: { ObjectId }
+} = mongoose;
+
+const generateComments = listData => {
+  const comments = [];
```
Comment (adamklepacz) [inline server/seed/demoSeed/generateComments.js]: add break line

File: server/seed/demoSeed/generateComments.js [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+  for (let i = 0; i < listData.length; i += 1) {
```
Comment (adamklepacz) [inline server/seed/demoSeed/generateComments.js]: store this lenght in a constanst before loop
I've read that is better fore loop perfomance. Cause it's cached above the loop and it doesn't have to be evaluated in every loop iteration.

File: server/seed/demoSeed/generateUsers.js [hunk 25]
```diff
@@ -0,0 +1,55 @@
+const mongoose = require('mongoose');
+
+const {
+  Types: { ObjectId }
+} = mongoose;
+const idFromProvider = process.env.DEMO_USER_ID_FROM_PROVIDER;
```
Comment (adamklepacz) [inline server/seed/demoSeed/generateUsers.js]: destructure and map to a name you want

File: server/seed/demoSeed/seedCohorts.js [hunk 26]
```diff
@@ -0,0 +1,12 @@
+const Cohort = require('../../models/cohort.model');
+const { generateCohorts } = require('./generateCohorts');
+
+const seedCohorts = async (demoUserId, userIds) => {
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedCohorts.js]: is this async needed?
I dont see any await inside

File: server/seed/demoSeed/seedCohorts.js [hunk 27]
```diff
@@ -0,0 +1,12 @@
...
+  const cohorts = generateCohorts(demoUserId, userIds);
+
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedCohorts.js]: remove break line

File: server/seed/demoSeed/seedComments.js [hunk 28]
```diff
@@ -0,0 +1,12 @@
+const Comment = require('../../models/comment.model');
+const { generateComments } = require('./generateComments');
+
+const seedComments = async listData => {
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedComments.js]: is this async needed?

File: server/seed/demoSeed/seedComments.js [hunk 29]
```diff
@@ -0,0 +1,12 @@
...
+  const comments = generateComments(listData);
+
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedComments.js]: remove break line

File: server/seed/demoSeed/seedDemoData.js [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+const { seedLists } = require('./seedLists');
+const { seedComments } = require('./seedComments');
+
+const seedDemoData = demoUserId => {
+  let cohortIds;
+  let userIds;
+
+  seedUsers(demoUserId)
+    .then(users => {
+      userIds = users.map(user => user._id);
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedDemoData.js]: always add break line before RETURN 

line 17, line 22, ETC...

File: server/seed/demoSeed/seedLists.js [hunk 31]
```diff
@@ -0,0 +1,12 @@
+const List = require('../../models/list.model');
+const { generateLists } = require('./generateLists');
+
+const seedLists = async (demoUserId, userIds, cohortIds) => {
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedLists.js]: is this async needed?

File: server/seed/demoSeed/seedLists.js [hunk 32]
```diff
@@ -0,0 +1,12 @@
...
+  const lists = generateLists(demoUserId, userIds, cohortIds);
+
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedLists.js]: remove break line ??

File: server/seed/demoSeed/seedUsers.js [hunk 33]
```diff
@@ -0,0 +1,12 @@
+const User = require('../../models/user.model');
+const { generateUsers } = require('./generateUsers');
+
+const seedUsers = demoUserId => {
+  const users = generateUsers(demoUserId);
+
```
Comment (adamklepacz) [inline server/seed/demoSeed/seedUsers.js]: remove break line

File: server/tasks/clearDemoData.js [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+
+const disconnectDatabase = () =>
+  mongoose
+    .disconnect()
+    .then(() => console.log('Successfully disconnected...'))
+    .catch(() => process.exit(1));
+
+const clearDemoData = () => {
+  const idFromProvider = process.env.DEMO_USER_ID_FROM_PROVIDER;
+  const mainDemoUserId = process.env.DEMO_USER_ID;
```
Comment (adamklepacz) [inline server/tasks/clearDemoData.js]: destructure from process and map to names you wanted

File: client/src/modules/authorization/AuthBox.jsx [hunk 35]
```diff
@@ -1,26 +1,44 @@
...
 
+  handleLaunchingDemo = () => {
+    const { loginDemoUser } = this.props;
+
+    return loginDemoUser();
+  };
+
   handleCookieAccept = () => this.setState({ isCookieSet: true });
 
+  handleDisablingLink = event => event.preventDefault();
```
[the remark this answers is not in the document]
Comment (alemikolo) [reply client/src/modules/authorization/AuthBox.jsx]: I removed this line - this handler was unused

File: client/src/modules/authorization/AuthBox.jsx [hunk 36]
```diff
@@ -36,19 +54,41 @@ class AuthBox extends PureComponent {
...
-                  className="google-button__img"
-                  src={GoogleButtonImg}
-                />
-              </a>
+              <div className="authbox__button-wrapper">
+                <a
+                  className={classNames('google-button', {
+                    'disabled-google-button': !isCookieSet || pending
+                  })}
+                  onClick={() => this.setState({ pending: true })}
```
[the remark this answers is not in the document]
Comment (alemikolo) [reply client/src/modules/authorization/AuthBox.jsx]: Yes, you're - It was already moved to separate handler method - I don't understand why some changes are not included here:
```
<a
  className={classNames('google-button', {
    'disabled-google-button': !isCookieSet || pending
  })}
  onClick={this.handleLogin}
  href="/auth/google"
  tabIndex={!isCookieSet ? '-1' : '1'}
>
```

File: server/common/utils/userUtils.js (hunk 11 above)
Comment (alemikolo) [reply server/common/utils/userUtils.js]: We should check ids in one array, but only in viewersIds. The demo user gets few demo lists, sometimes they are an owner, sometimes a member, sometimes a viewer. We want to remove all list where demo user is a viewer.

**Comment by alemikolo** [comment]
To make this functionality working we have to
1. Save in .env file to constants: DEMO_USER_ID, DEMO_USER_ID_FROM_PROVIDER.
If there is no demo user in DB, app will create a new demo user using those constants. DEMO USER_ID is 24 chars string which will be used to create Id - we can't generate it automatically. DEMO_USER_ID_FROM_PROVIDER is a string which will be saved as an IdFromProvider and will be the same for all demo users.
2. Remove index from the "idFromProvider field" in the users collection in DB.

**Comment by adamklepacz** [comment]
> To make this functionality working we have to:
> 
> 1. Save in .env file to constants: DEMO_USER_ID, DEMO_USER_ID_FROM_PROVIDER.
>    If there is no demo user in DB, app will create a new demo user using those constants. DEMO USER_ID is 24 chars string which will be used to create Id - we can't generate it automatically. DEMO_USER_ID_FROM_PROVIDER is a string which will be saved as an IdFromProvider and will be the same for all demo users.
> 2. Remove index from the "idFromProvider" field in the users collection in DB.

Add this info to the readme file in a short understandable form

File: server/controllers/list.js [hunk 37]
```diff
@@ -751,14 +751,21 @@ const removeMemberRole = (req, resp) => {
...
   const { email } = req.body;
+  const { DEMO_MODE_ID } = process.env;
   let list;
   let user;
   let cohortMembers = [];
 
+  if (idFromProvider === DEMO_MODE_ID) {
+    return resp
+      .status(401)
+      .send({ message: 'Adding members is disabled for demo purposes.' });
```
Comment (marekrozmus) [inline server/controllers/list.js]: Please change 'Adding members is disabled in demo mode'

**Pull request merged**